keps/sig-api-machinery/2309-idl/README.md

```diff
@@ -0,0 +1,609 @@
+
+* Writing Go code to generate CRDs makes little sense in environments
+  without any pre-existing Go code
+
+* Generation is in-binary, making custom extensions (even those in Go)
+  more difficult to write without requiring duplicate parsing passes
+
+#### Wire formats are strongly tied to source formats
+
+Because we rely on Go types to serialize to JSON, experimenting with
+different structures in Go is difficult (e.g. it might rely on
+hand-writing custom JSON serialization)
+
+While it's not *remotely* in scope for this proposal to change the Go
+types representation, it has occasionally come up that it'd be nice to
+experiment (e.g. types the preserve unknown fields round-trip, interfaces
+instead of giant structs where you have to set only one field).
+
+#### Pseudo-IDL is hard to review
+
+* For many of the reasons pseudo-IDL is [hard to
+  write](#pseudo-idl-is-hard-to-write-properly), it's also hard to review:
+  reviewers have to look at many different pieces of information, commonly
+  with slightly different values.  This results in issues like [bad struct
+  tags][upper-field-name].
+
+* Different parts of the pseudo-IDL are split across different files
+  (types vs validation vs defaulting), making it more difficult to review
+  wholistically.
+
+### ...as tooling maintainers?
+
+* Duplicated efforts: [k8s.io/code-generator][code-generator] and
+  [controller-gen][controller-tools] use different tooling to generate
+  code & artifacts from markers.  This results in:
+
+  * duplicated maintenance effort
+  * slightly differing semantics
+  * generation modules written in the less flexible k8s.io/code-generator
+    cannot easily being incompatible with controller-gen
+
+* Difficulty parsing:
+
+  * a full Go parser & type-checker are required (due to aliasing over
+    primitives, dot imports, etc), as well as a full Go standard library
+
+  * Parsing is slow, since it requires traversing non-IDL resources in the
+    same package (hacks can speed this up, but only so much, and at the
+    expense of good error messages)
+
+  * Properly parsing markers is clumsy & easy to mess up between different
+    implementations, since it relies on wobbly definitions of which comments
+    are associated with a given field or type
+
+  * It's hard to distribute your pseudo-IDL without distributing the rest
+    of your types package (this is a problem for orgs that have
+    restrictions on which source code they can deliver to end users)
+
+* Pseudo-IDL is hard to validate: since markers are just comment, and are
+  explicitly defined, namespaced, etc, it's hard to know if
+  a non-conforming or unknown marker is actually wrong, or just intended
+  to be consumed by a different piece of tooling, or just happens to look
+  like a marker.
+
+## Goals
+
+1. **Guide authors in the right direction**: it should be easy to write
+   APIs that follow the Kubernetes API conventions and recommendations,
+   and hard/impossible to write invalid APIs
```

Comment:
Who are the consumers of this goal?

I assume you will advocate that IDL be used within Kube - are you proposing new kube apis use this mechanism (and thus the IDL constrains what apis can be expressed)?

For a non-core apis author, I assume this is purely optional, so this won’t prevent someone writing a kube api the idl can’t represent?

The goals kind of need to have the anticipated consumer identified - it’s hard to say whether these are good goals without a clearer idea of how this is useful (vs the simpler options or doing nothing).

Reply:
> Who are the consumers of this goal?

I'm happy to clarify this a bit, but I feel like this applies to all the anticipate consumers I've outlined in the "who uses API types" section.

> I assume you will advocate that IDL be used within Kube - are you proposing new kube apis use this mechanism (and thus the IDL constrains what apis can be expressed)?

Yes, new & existing APIs in k/k will be ported to the IDL.  The goal is to be able to express all of k/k, at the very least.  I've run the converter on all of core/v1 to get a bit of a signal (and it works). 

> For a non-core apis author, I assume this is purely optional, so this won’t prevent someone writing a kube api the idl can’t represent?

In depth discussion below, but TL;DR: I think there's value in guiding community contributors to use our rules, but it's a non-goal that this replace OpenAPI in CRDs, for instance, so nothing's stopping you from writing that.

I'd be particularly interested, though, to see if there's particular APIs that you're concerned about being expressible.

---

Particularly, as one of the primary maintainers of KubeBuilder & friends, I've been on the "supporting" end of people writing all sorts of different API, and then coming back and saying "hey, thing X doesn't quite work" (e.g. floats) or "I designed my API with thing X, and now I can't use new CRD versions & features" (e.g. complex oneof, anyof, not, etc).

If we're gonna say "hey, you can't do this thing in core k/k APIs", either there's a good reason and we should "communicate" that to the broader community via our language, or we should not make that rule.

I certainly don't want to be the bearer of bad news of "hey, we put a lot of time & effort into this API and now we can't use it with CRD v1/SSA/whatever".

FWIW, we already "soft" constrain what APIs people can write, without communicating it clearly (structural schema, for instance), and it's caused a *lot* of confusion in the end user community.

All that being said: notice that it's explicitly a non-goal to replace openapi everywhere.  Nothing's stopping you from writing OpenAPI by hand (besides, you know, writing OpenAPI by hand & our arcane restrictions on what's allowed) -- some stuff just might not be expressible in the IDL.

Reply:
FYI, I've added a section ("Concrete KubeBuilder & k/k Changes") with the intentions specified in the comment above

Reply:
If you can't convince everyone to use the IDL (the new minimum expressible subset) you can't remove support for all the other ways of doing this (openapi, CRD, validation) so you now have 4 problems :)  That's the key challenge I'm trying to get you to articulate - what angle helps drive the change you seek with the right audience so the problem goes away.  I would argue that the goal is to clearly identify which minimal subset is the authoritative one, and not allow for others to define a different, distinct, minimal subset.  Which if we said was OpenAPI would then mean the IDL was only useful for some people some of the time.

Reply:
> If you can't convince everyone to use the IDL (the new minimum expressible subset) you can't remove support for all the other ways of doing this (openapi, CRD, validation) so you now have 4 problems :) 

There is merit to code review even if it isn't so comprehensive that one could just delete their test suites.

Reply:
exactly what @thockin said.  It wouldn't surprise me if there was a type out there somewhere in third-party land that technically half-worked but wildly violated all of our best practices and was poised to topple over at any point in time, but base on my efforts supporting CT, my aim is to cover almost all of the third-party types, and *definitely* cover all k/k types.